src/msg/async/frames_v2.cc

```diff
@@ -0,0 +1,191 @@
+                   sizeof(preamble));
+
+  epilogue_plain_block_t epilogue;
+  // FIPS zeroization audit 20191115: this memset is not security related.
+  ::memset(&epilogue, 0, sizeof(epilogue));
+
+  for (size_t i = 0; i < m_descs.size(); i++) {
+    ceph_assert(segment_bls[i].length() == m_descs[i].logical_len);
+    epilogue.crc_values[i] = segment_bls[i].crc32c(-1);
+    if (segment_bls[i].length() > 0) {
+      frame_bl->claim_append(segment_bls[i]);
+    }
+  }
+  frame_bl->append(reinterpret_cast<const char*>(&epilogue),
+                   sizeof(epilogue));
+}
+
+void FrameAssembler::asm_secure_r0(const preamble_block_t& preamble,
+                                   bufferlist* segment_bls,
+                                   bufferlist* frame_bl) const {
+  bufferlist preamble_bl;
+  preamble_bl.append(reinterpret_cast<const char*>(&preamble),
+                     sizeof(preamble));
+
+  epilogue_secure_block_t epilogue;
+  // FIPS zeroization audit 20191115: this memset is not security related.
+  ::memset(&epilogue, 0, sizeof(epilogue));
+  bufferlist epilogue_bl;
+  epilogue_bl.append(reinterpret_cast<const char*>(&epilogue),
+                     sizeof(epilogue));
+
+  // preamble + MAX_NUM_SEGMENTS + epilogue
+  uint32_t onwire_lens[MAX_NUM_SEGMENTS + 2];
+  onwire_lens[0] = preamble_bl.length();
+  for (size_t i = 0; i < m_descs.size(); i++) {
+    onwire_lens[i + 1] = segment_bls[i].length();  // already padded
+  }
+  onwire_lens[m_descs.size() + 1] = epilogue_bl.length();
+  m_crypto->tx->reset_tx_handler(onwire_lens,
+                                 onwire_lens + m_descs.size() + 2);
+  m_crypto->tx->authenticated_encrypt_update(preamble_bl);
+  for (size_t i = 0; i < m_descs.size(); i++) {
+    if (segment_bls[i].length() > 0) {
+      m_crypto->tx->authenticated_encrypt_update(segment_bls[i]);
+    }
+  }
+  m_crypto->tx->authenticated_encrypt_update(epilogue_bl);
+  append_encrypt_final(frame_bl);
+}
+
+bufferlist FrameAssembler::assemble_frame(Tag tag, bufferlist* segment_bls,
+                                          const uint16_t* segment_aligns,
+                                          size_t segment_count) {
+  m_descs.resize(calc_num_segments(segment_bls, segment_count));
+  for (size_t i = 0; i < m_descs.size(); i++) {
+    m_descs[i].logical_len = segment_bls[i].length();
+    m_descs[i].align = segment_aligns[i];
+  }
+
+  preamble_block_t preamble;
+  fill_preamble(tag, &preamble);
+
+  bufferlist frame_bl;
+  if (m_crypto->rx) {
+    for (size_t i = 0; i < m_descs.size(); i++) {
+      ceph_assert(segment_bls[i].length() == m_descs[i].logical_len);
+      // We're padding segments to biggest cipher's block size. Although
+      // AES-GCM can live without that as it's a stream cipher, we don't
+      // want to be fixed to stream ciphers only.
+      pad_zero(&segment_bls[i], get_segment_padded_len(i));
+    }
+    asm_secure_r0(preamble, segment_bls, &frame_bl);
```

Comment:
I think we would incur a `bufferlist` move in that case?

Reply:
I think you would have gotten NRVO. Nowadays compilers are pretty good at eliding copies/moves and [C++17 mandated a bunch of cases previously considered  as *allowances*](https://en.cppreference.com/w/cpp/language/copy_elision).

Reply:
I don't think the compiler can apply (N)RVO if the calling function stores into the existing variable:
```
bufferlist bl = make_bl(...);  // (N)RVO or move assignment, depending on make_bl()
```
```
bufferlist bl;
bl = make_bl(...);  // move assignment
```
Here it would be the latter, because `frame_bl` must be assigned to in four different branches before the common assert.

I didn't think about moves or attempt to micro-optimize here, I wrote that naturally.  I do a lot of work in C, perhaps it shows ;)

Reply:
I think problem in the second arrangement isn't about applying NRVO per se. IIUC what happens there is:

1. Default initialization of `bl`.
2. The call and fabrication where a bunch of temporaries could be knocked-out with (N)RVO.
3. Move assignment.

There is no rule allowing compiler to disregard visible side effects of 1. and 3. This was pretty problematic when e.g. returning tuple (multiple values) and using `std::tie` to unpack it. C++17 addressed that with structured bindings, BTW.

> Here it would be the latter, because `frame_bl` must be assigned to in four different branches before the common assert.

I see. This wouldn't be one-liner. :-( It would require extracting the assertion to `Frame::get_buffer()` or introducing intermediary function.